src/ipc/Queue.h

```diff
@@ -410,6 +426,96 @@ OneToOneUniQueue::push(const Value &value, QueueReader *const reader)
+{
+    // Nobody can modify our theIn so, after capturing some valid theSize value
+    // in count, we can reliably report all [theIn-count, theIn) items that were
+    // queued at theSize capturing time. We may report items already pop()ed by
+    // the other side, but that is OK because pop() does not modify items -- it
+    // only increments theOut.
+    const auto count = theSize.load();
+    statOpen(entry, "pushIndex", "other", count);
+    statClose<Value>(entry, theIn - count, count); // Branch XXX: theIn-count may get negative!
+}
+
+template <class Value>
+void
+OneToOneUniQueue::statOut(StoreEntry &entry) const
+{
+    // Nobody can modify our theOut so, after capturing some valid theSize value
+    // in count, we can reliably report all [theOut, theOut+count) items that
+    // were queued at theSize capturing time. We will miss new items push()ed by
+    // the other side, but that is OK.
+    const auto count = theSize.load();
+    statOpen(entry, "other", "popIndex", count);
+    statClose<Value>(entry, theOut, count);
+}
+
+/// start cache manager reporting (by reporting queue parameters)
+/// The labels reflect whether the caller owns theIn or theOut data member and,
+/// hence, can report the corresponding value reliably.
+inline void
+OneToOneUniQueue::statOpen(StoreEntry &entry, const char *inLabel, const char *outLabel, const uint32_t count) const
+{
+    // Branch XXX: Some arguments are not %u (i.e. unsigned int)!
+    storeAppendPrintf(&entry, "{ size: %u, capacity: %u, %s: %u, %s: %u",
+                      count, theCapacity, inLabel, theIn, outLabel, theOut);
+}
+
+/// end cache manager reporting with a sample of [start, start + size) items
+template <class Value>
+void
+OneToOneUniQueue::statClose(StoreEntry &entry, const unsigned int start, const uint32_t count) const
+{
+    if (!empty()) {
+        storeAppendPrintf(&entry, ", items: [\n");
+        // report a few leading and trailing items, without repetitions
+        const auto sampleSize = std::min(3U, count); // leading/trailing sample
+        statRange<Value>(entry, start, 0, sampleSize);
+        if (sampleSize < count) { // the first sample did not show some items
+            if (sampleSize*2U < count) { // both samples will not show some items
+                const auto middleCount = count - sampleSize*2U;
+                if (middleCount == 1)
+                    statRange<Value>(entry, start, sampleSize, 1);
+                else
+                    storeAppendPrintf(&entry, "    # ... %u items not shown ...\n", middleCount);
+            }
+            // The `start` offset aside, the first sample reported all items
+            // below the sampleSize offset. The second sample needs to report
+            // the last sampleSize items (i.e. starting at count-sampleSize
+            // offset) except those already reported by the first sample.
+            const auto secondSampleOffset = std::max(sampleSize, count - sampleSize);
```

Comment:
Here is a bug: we need 'min' instead of 'max'. I'll fix that.

Reply:
Yes, max is correct. I suspect that the code would be simpler if statRange() is adjusted to ignore/skip out-of-range offsets. However, I did not have enough time to express that idea in code.

Reply:
Done at be21947.